### PR TITLE
Add a script to automate product addition and configuration

### LINKED ISSUE #1401: Create a script to make new iso available
**Component**:

'salt'

**Why this is needed**:

Currently the process to make new ISOs available is really not user friendly

**What should be done**:

Create a script to simplify all this process and update the doc accordingly 
https://github.com/scality/metalk8s/blob/development/2.0/docs/operation/preparation.rst

**Implementation proposal** (strongly recommended):

Create a bash script (like the bootstrap.sh one) that will:
- take a path to ISO(s) (or a directory(ies)) as argument
- add these path(s) to the bootstrap.yaml config file
- execute the state to moiunt all iso
- **for each versions** execute the state (local mode) to configure salt-master and repositories
- execute the state to make all the versions available

This script should:
- only output few lines (like bootstrap script)
- have a verbose mode (like bootstrap script)
- have a log file (like bootstrap script)
- be runnable without any agurments to just reconfigure all repo(s) and salt-master
- be able to take multiple ISO(s) ∕ directories as arguments
- have a dry run mode (maybe only running the states with `test=True`)

Update the operation guide documentation

### PR DESCRIPTION
**Pull request by sayf-eddine-scality**
Fixes: #1401

**Component**: salt, upgrade, downgrade

**Context**: 
The steps to add a product and configure it is very manual, we need a script to automate it
**Summary**:
A script has been added that automate product configurations
**Acceptance criteria**: 
Check the script help and use it to configure a new metalk8s iso

---

Closes: #1401

### CODE REVIEW

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

File: scripts/iso-manager.sh [hunk 1]
```diff
@@ -0,0 +1,149 @@
+#!/bin/bash
+
+BOOTSTRAP_CONFIG="/etc/metalk8s/bootstrap.yaml"
+DEBUG="info"
+LOGFILE="/tmp/iso.log"
+DRY_RUN="False"
+
+# Helpers
+_usage() {
+echo "iso-manager.sh [options]"
```
Comment (slaperche-scality) [inline scripts/iso-manager.sh]: The function body should be indented.

File: scripts/iso-manager.sh [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+echo "Options:"
+echo "-p/--product:     Path to product folder or ISO"
+echo "--d/--dry-run:    Run actions in dry run mode"
```
Comment (slaperche-scality) [inline scripts/iso-manager.sh]: single hyphen for short option `-d` not `--d`

File: scripts/iso-manager.sh [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        "$@" >> "$LOGFILE" 2>&1
+    else
+        "$@" >&2
+    fi
+}
+
+# Run commands and display/write the output to stdout/logfile
+_info() {
+    if [ "$DEBUG" == "info" ]; then
+        if [ -n "$LOGFILE" ]; then
```
Comment (slaperche-scality) [inline scripts/iso-manager.sh]: l.35 to 39 could be replace by a call to `_log`, no?

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

File: scripts/iso-manager.sh [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  for e in "$@"; do
+      [[ "$e" == "$match" ]] && return 0;
+  done
+  return 1
+}
+
+_add_products() {
+    # Use salt file.serialize merge require having full list
+    products=($(salt-call pillar.get metalk8s:products \
+        --out txt | cut -d' ' -f2- | tr -d '[],'))
```
Comment (slaperche-scality) [inline scripts/iso-manager.sh]: Having an output example in comment would help to understand the "parsing" done.

File: scripts/iso-manager.sh [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    for p in "$@"; do
```
Comment (slaperche-scality) [inline scripts/iso-manager.sh]: `product` instead of `p`

File: scripts/iso-manager.sh [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    if [ -z "$SALTENV" ]; then
+        SALTENV="metalk8s-$(salt-call --out txt slsutil.renderer \
+            string="{{ pillar.metalk8s.nodes[grains.id].version }}" \
+            | cut -c 8-)"
+    fi
+}
+
+# helper function to check for element in array
+containsElement () {
+  local e match="$1"
```
Comment (slaperche-scality) [inline scripts/iso-manager.sh]: `item`, `element`, … instead of `e`

File: scripts/iso-manager.sh [hunk 7]
```diff
@@ -0,0 +1,154 @@
+#!/bin/bash
```
Comment (slaperche-scality) [inline scripts/iso-manager.sh:1]: A little more info in the commit message would be welcome (what the script is doing for instance: by reading the code it seems to auto-mount stuff and doing things after)

**Review by slaperche-scality — DISMISSED** [review]
I'm not a Shell guru, so if someone else could take a look that would be great.

File: scripts/iso-manager.sh [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+LOGFILE="/tmp/iso.log"
+DRY_RUN="False"
+
+# Helpers
+_usage() {
+echo "iso-manager.sh [options]"
+echo "Options:"
+echo "-p/--product:     Path to product folder or ISO"
+echo "-d/--dry-run:     Run actions in dry run mode"
+echo "-l/--log-file:    path to log file"
```
Comment (TeddyAndrieux) [inline scripts/iso-manager.sh]: To be consistent use `verbose` instead of `DEBUG` and add an option to change it

File: scripts/iso-manager.sh [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+            $(salt-call --out txt pillar.get metalk8s:endpoints | cut -c 8-)}}" \
+            test=$DRY_RUN
+    done
+    # Make the new version available
+    _log echo "Making new versions available"
+    _info salt-call state.sls metalk8s.products saltenv="$SALTENV" test="$DRY_RUN"
+}
+
+# Main
+[ $# -lt 1 ] && _usage
```
Comment (TeddyAndrieux) [inline scripts/iso-manager.sh]: Like described in the ticket, this script should support no argument and will basically just skip `add_products`

**Review by TeddyAndrieux — DISMISSED** [review]

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

**Review by TeddyAndrieux — COMMENTED** [review]
The script is not included in the ISO :(

Also you need to update the doc according to the process change

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

**Comment by sayf-eddine-scality** [comment]
@bert-e approve

File: buildchain/buildchain/iso.py [hunk 10]
```diff
@@ -115,6 +116,23 @@ def task__iso_add_node_manifest() -> types.TaskDict:
...
+    iso_manager = [
+            constants.ROOT/'scripts'/'iso-manager.sh',
+            dest
+            ]
+    return {
+            'title': lambda task: utils.title_with_target1('GENERATE', task),
+            'actions': [(coreutils.cp_file, iso_manager)],
+            'targets': [dest],
+            'task_dep': ['_iso_mkdir_root'],
+            'uptodate': [True],
```
Comment (slaperche-scality) [inline buildchain/buildchain/iso.py]: Nope.
Why do you want it to be always true?

If the script is modified, you want to copy it again.
You should have `constants.ROOT/'scripts'/'iso-manager.sh'` as `file_dep`.

File: buildchain/buildchain/iso.py [hunk 11]
```diff
@@ -115,6 +116,23 @@ def task__iso_add_node_manifest() -> types.TaskDict:
...
 
+def task__iso_add_iso_manager() -> types.TaskDict:
+    """Copy the iso manager script to scripts."""
+    dest = constants.ISO_ROOT/'iso-manager.sh'
+    iso_manager = [
+            constants.ROOT/'scripts'/'iso-manager.sh',
+            dest
+            ]
+    return {
+            'title': lambda task: utils.title_with_target1('GENERATE', task),
```
Comment (slaperche-scality) [inline buildchain/buildchain/iso.py]: You're not generating anything, just copying a file, so "COPY" sounds better.

File: buildchain/buildchain/iso.py (hunk 10 above)
Comment (slaperche-scality) [reply buildchain/buildchain/iso.py]: And looking at the code, it seems like `task__iso_add_node_manifest` was wrong as well and should be fixed (in another commit/PR, not here).

File: buildchain/buildchain/iso.py [hunk 12]
```diff
@@ -115,6 +116,23 @@ def task__iso_add_node_manifest() -> types.TaskDict:
      }
 
 
+def task__iso_add_iso_manager() -> types.TaskDict:
+    """Copy the iso manager script to scripts."""
+    dest = constants.ISO_ROOT/'iso-manager.sh'
+    iso_manager = [
```
Comment (slaperche-scality) [inline buildchain/buildchain/iso.py]: Weird layout.

Should be:

```
iso_manager = [
    constants.ROOT/'scripts'/'iso-manager.sh',
    dest
]
````

**Review by slaperche-scality — CHANGES_REQUESTED** [review]

File: docs/operation/preparation.rst [hunk 13]
```diff
@@ -3,43 +3,10 @@ ISO Preparation
...
-
-#. Mount the ISO file.
-
-   .. code::
-
-      salt-call state.sls metalk8s.products.mounted saltenv=metalk8s-$(salt-call --out txt slsutil.renderer string="{{ pillar.metalk8s.nodes[grains.id].version }}" | cut -c 8-)
-
-#. Configure the new repositories and salt-master.
-   Replace **<destination_version>** with the appropriate version number.
+To provision a new **Metalk8s** ISO you need to run the according utility
```
Comment (slaperche-scality) [inline docs/operation/preparation.rst]: remove `accordingly`

**Review by slaperche-scality — DISMISSED** [review]

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

**Review by slaperche-scality — DISMISSED** [review]

**Comment by sayf-eddine-scality** [comment]
@bert-e reset

**Review by slaperche-scality — APPROVED** [review]

**Pull request merged**